src/Storages/ExternalStream/Kafka/Kafka.cpp

```diff
@@ -276,45 +363,185 @@ void Kafka::validateMessageKey(const String & message_key_, IStorage * storage,
         {
-            if (shard >= shards)
+            if (shard >= parition_count)
                 throw Exception(
                     ErrorCodes::INVALID_SETTING_VALUE,
-                    "Invalid topic partition {} for topic {}, biggest partition is {}",
+                    "Invalid topic partition {} for topic {}, total number of partitions is {}",
                     shard,
                     settings->topic.value,
-                    shards);
+                    parition_count);
         }
     }
 }
 
+Pipe Kafka::read(
+    const Names & column_names,
+    const StorageSnapshotPtr & storage_snapshot,
+    SelectQueryInfo & query_info,
+    ContextPtr context,
+    QueryProcessingStage::Enum /*processed_stage*/,
+    size_t max_block_size,
+    size_t /*num_streams*/)
+{
+    auto consumer = getConsumer();
+    /// The topic_ptr can be shared between all the sources in the same pipe, because each source reads from a different partition.
+    auto topic_ptr = std::make_shared<RdKafka::Topic>(*consumer->getHandle(), topic());
+
+    /// User can explicitly consume specific kafka partitions by specifying `shards=` setting
+    /// `SELECT * FROM kafka_stream SETTINGS shards=0,3`
+    std::vector<int32_t> shards_to_query;
+    if (!shards_from_settings.empty())
+        shards_to_query = shards_from_settings;
+    else
+    {
+        /// The number of paritions could have been changed, always query the number before
+        auto parition_count = topic_ptr->describe();
+
+        /// Query all shards / partitions
+        shards_to_query.reserve(parition_count);
+        for (int32_t i = 0; i < parition_count; ++i)
+            shards_to_query.push_back(i);
+    }
+    LOG_INFO(logger, "Reading from partitions [{}] of topic {}", fmt::join(shards_to_query, ","), topic());
+
+    assert(!shards_to_query.empty());
+
+    Pipes pipes;
+    pipes.reserve(shards_to_query.size());
+
+    // const auto & settings_ref = context->getSettingsRef();
+    /*auto share_resource_group = (settings_ref.query_resource_group.value == "shared") && (settings_ref.seek_to.value == "latest");
+    if (share_resource_group)
+    {
+        for (Int32 i = 0; i < shards; ++i)
+        {
+            if (!column_names.empty())
+                pipes.emplace_back(source_multiplexers->createChannel(i, column_names, metadata_snapshot, context));
+            else
+                pipes.emplace_back(source_multiplexers->createChannel(i, {RESERVED_APPEND_TIME}, metadata_snapshot, context));
+        }
+    }
+    else*/
+    {
+        /// For queries like `SELECT count(*) FROM tumble(table, now(), 5s) GROUP BY window_end` don't have required column from table.
+        /// We will need add one
+        Block header;
+        if (!column_names.empty())
+            header = storage_snapshot->getSampleBlockForColumns(column_names);
+        else
+            header = storage_snapshot->getSampleBlockForColumns({ProtonConsts::RESERVED_EVENT_TIME});
```

Comment:
The external stream has not `_tp_time` , we can use first column

Reply:
External stream has `_tp_time`, it's the messages' timestamps.